### PR TITLE
[FEATURE]: Add methods for translating multiple ADF pipelines

### LINKED ISSUE #7: [FEATURE]: Add methods for translating multiple ADF pipelines
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Problem statement

wkmigrate's definition stores allow loading and translating a single ADF pipeline. Customers would like a simpler experience for translating many pipelines at once.

### Proposed Solution

Add methods to load and translate multuple ADF pipelines at once.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by ghanse**
## Changes

Add batch methods for loading and translating multiple ADF pipelines at once, addressing customer requests for a simpler multi-pipeline migration experience.

**FactoryDefinitionStore:**
- `list_pipelines()` — returns names of all pipelines in the Data Factory
- `load_all(pipeline_names=None)` — loads and translates multiple pipelines; defaults to all pipelines when no names are given; gracefully skips individual failures

**FactoryClient:**
- `list_pipelines()` — lists pipeline names via the ADF management API

**WorkspaceDefinitionStore:**
- `to_jobs(pipeline_definitions)` — creates a Databricks job for each pipeline
- `to_asset_bundles(pipeline_definitions, bundle_directory)` — writes an asset bundle per pipeline into subdirectories

### Linked issues

Resolves #7

### Tests

- [ ] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] added end-to-end tests
- [ ] added performance tests

9 new tests in `tests/test_batch_translation.py` covering:
- `list_pipelines` returns all fixture pipeline names
- `load_all` with explicit names, default-all, skip-on-failure, all-fail edge case
- `to_jobs` with multiple pipelines and empty list
- `to_asset_bundles` creates per-pipeline subdirectories and handles empty list

All 241 tests pass.

### CODE REVIEW

File: src/wkmigrate/definition_stores/workspace_definition_store.py [hunk 1]
```diff
@@ -92,6 +92,43 @@ def __post_init__(self) -> None:
...
+            pipeline_definitions: List of ``Pipeline`` dataclasses to deploy.
+
+        Returns:
+            List of job identifiers registered in the workspace.
+        """
+        job_ids: list[int] = []
+        for pipeline_definition in pipeline_definitions:
+            job_id = self.to_job(pipeline_definition)
+            if job_id is not None:
+                job_ids.append(job_id)
```
Comment (ghanse) [inline src/wkmigrate/definition_stores/workspace_definition_store.py]: **Suggestion: `to_jobs` silently drops pipelines where `to_job` returns `None`.**

`to_job` returns `None` when job creation fails (via the `raise ValueError` path this shouldn't actually happen, but the return type is `int | None`). Here you silently skip `None` results. For consistency with `load_all` — which logs failures — consider logging a warning when `to_job` returns `None` so callers know which pipelines were not deployed. Otherwise the caller gets back fewer job IDs than pipelines they passed in, with no indication of what went wrong.

```python
for pipeline_definition in pipeline_definitions:
    job_id = self.to_job(pipeline_definition)
    if job_id is not None:
        job_ids.append(job_id)
    else:
        logger.warning("to_job returned None for pipeline '%s', skipping", pipeline_definition.name)
```

Alternatively, consider whether `to_jobs` should also wrap each call in a try/except like `load_all` does, so one failed deployment doesn't abort the entire batch.

Comment (ghanse) [reply src/wkmigrate/definition_stores/workspace_definition_store.py]: We should wrap each call in try and except.

File: tests/test_batch_translation.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+from wkmigrate.definition_stores.workspace_definition_store import WorkspaceDefinitionStore
+from wkmigrate.models.ir.pipeline import (
+    DatabricksNotebookActivity,
+    Pipeline,
+)
+
+
+# ---------------------------------------------------------------------------
+# Helpers
+# ---------------------------------------------------------------------------
```
Comment (ghanse) [inline tests/test_batch_translation.py]: Remove inline comments.

File: src/wkmigrate/definition_stores/workspace_definition_store.py [hunk 3]
```diff
@@ -92,6 +92,43 @@ def __post_init__(self) -> None:
...
+        Each pipeline is written to a subdirectory named after the pipeline.
+
+        Args:
+            pipeline_definitions: List of ``Pipeline`` dataclasses to export.
+            bundle_directory: Parent directory for all generated bundles.
+            download_notebooks: If True, downloads referenced notebooks from the workspace.
+        """
+        for pipeline_definition in pipeline_definitions:
+            sub_directory = os.path.join(bundle_directory, pipeline_definition.name)
+            self.to_asset_bundle(pipeline_definition, sub_directory, download_notebooks=download_notebooks)
```
[the remark this answers is not in the document]
Comment (ghanse) [reply src/wkmigrate/definition_stores/workspace_definition_store.py]: Done. Each `to_asset_bundle` call is now wrapped in try/except with a warning log on failure.

File: tests/test_batch_translation.py (hunk 2 above)
Comment (ghanse) [reply tests/test_batch_translation.py]: Done. Removed inline section-header comments.

**Comment by ghanse** [comment]
## Review feedback addressed

All inline review comments from ghanse and Copilot have been addressed in commit 975288b:

1. **`to_jobs` — try/except with logging**: Each `to_job` call is now wrapped in try/except; failures are logged via `logger.warning` and skipped, consistent with `load_all`.
2. **`to_asset_bundles` — try/except with logging**: Each `to_asset_bundle` call is now wrapped in try/except with warning logging on failure.
3. **`to_asset_bundles` — pipeline name sanitization**: Non-alphanumeric characters in pipeline names are replaced with `_` using `re.sub` to prevent path traversal.
4. **`factory_client.list_pipelines` — removed dead None check**: The `management_client is None` guard was removed since `__post_init__` always initializes it.
5. **`conftest.py` — `os.path.join` for paths**: Replaced f-string path construction with `os.path.join` for portability.
6. **`conftest.py` — `mock_factory_store` fixture**: Added a shared pytest fixture to eliminate repeated `FactoryDefinitionStore` construction in tests.
7. **`test_batch_translation.py` — `caplog` assertion**: `test_load_all_skips_failing_pipelines` now asserts that `DOES_NOT_EXIST` appears in the warning log.
8. **`test_batch_translation.py` — removed inline comments**: All section-header comments removed per reviewer request.
9. **Docs regenerated**: Ran `make docs-clean && make docs` and committed the updated API reference.

All 241 tests pass.

File: tests/test_batch_translation.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        client_secret="SECRET",
+        subscription_id="SUBSCRIPTION_ID",
+        resource_group_name="RESOURCE_GROUP",
+        factory_name="FACTORY_NAME",
+    )
+
+    # Include a non-existent pipeline name alongside a valid one
+    pipelines = store.load_all(pipeline_names=["TEST_PIPELINE_NAME", "DOES_NOT_EXIST"])
+    assert len(pipelines) == 1
+    assert pipelines[0].name == "TEST_PIPELINE_NAME"
```
[the remark this answers is not in the document]
Comment (ghanse) [reply tests/test_batch_translation.py]: Done. The `caplog` assertion for `load_all` was already added. Also added analogous `caplog`-based failure-path tests for `to_jobs` and `to_asset_bundles`.

File: src/wkmigrate/definition_stores/workspace_definition_store.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ghanse) [reply src/wkmigrate/definition_stores/workspace_definition_store.py]: Already addressed in a prior commit — pipeline names are sanitized with `re.sub(r"[^A-Za-z0-9_]", "_", ...)` and each call is wrapped in try/except. Now also covered by `test_to_asset_bundles_skips_failing_pipeline`.

**Comment by ghanse** [comment]
## Review feedback addressed

Added a new **"Translate multiple pipelines at once"** section to the usage guide (`docs/wkmigrate/docs/guide.mdx`) covering:

- `list_pipelines` and `load_all` for discovering and batch-translating pipelines
- `to_asset_bundles` for generating per-pipeline Databricks Asset Bundles
- `to_jobs` for batch-deploying pipelines as Lakeflow Jobs
- Error-handling behavior (log and skip on failure)

Also ran `make docs-clean && make docs` to regenerate the Docusaurus site. All 243 tests pass.

All prior review comments were already addressed in earlier commits. Replied to the remaining unacknowledged "remove inline comments" threads.

**Comment by ghanse** [comment]
## Final Review (post-cleanup) — PR #29

**Verdict: No blocking issues. Ready for approval.**

### Cleanup verification

- **Sanitization regex**: `[^A-Za-z0-9_-]` correctly preserves hyphens — tested
- **Collision detection**: `_1`, `_2` suffix with `while` loop + `seen_names` set prevents directory overwrites — tested
- **Empty name guard**: Raises `ValueError` before filesystem operations — tested
- **Test coverage**: 15 new tests cover happy paths, empty inputs, failures, sanitization, hyphens, collisions, empty names

### What looks good

- **Issue #7 scope**: All batch methods implemented — `list_pipelines`, `load_all`, `to_jobs`, `to_asset_bundles`
- **Error handling**: Consistent try/except + `logger.warning(exc_info=True)` + skip across all batch methods
- **Path safety**: Sanitization + collision detection + `os.path.abspath` prevents path traversal and overwrites
- **Test coverage**: 274 tests passing, thorough coverage of edge cases
- **Documentation**: Guide section added, API docs regenerated

### Non-blocking (future work)

Triple-collision scenario (`["foo", "foo", "foo_1"]`) produces `["foo", "foo_1", "foo_1_1"]` — the third pipeline's natural name gets displaced. Correct behavior (no clash) but could be surprising. Address if users report.

Transitioning to `needs-approval`.
